feat: add static shortcuts setup (resource templates, docs, one-shot shortcut extras)

MaximBelov opened the pull request:
## Summary

Adds first-class support for **static (manifest-declared) shortcuts** alongside the dynamic ones the plugin already handles:

- ready-to-copy Android resource templates under `resources/android/shortcuts/`,
- a README section explaining how to wire them up via `config.xml`,
- an Ionic / Angular usage example based on the `@awesome-cordova-plugins` wrapper,
- a small fix in `ShortcutsPlugin.getIntent()` so a shortcut is only reported once.

Static shortcuts need no JS API — they are declared in `res/xml/shortcuts.xml` and referenced from the manifest — but there was nothing in the repo showing how to do that with Cordova's `<resource-file>` / `<edit-config>`, nor how to distinguish a shortcut launch from a normal one.

## What's included

### 1. Resource templates (`resources/android/shortcuts/`)

| File | Copied to | Purpose |
| --- | --- | --- |
| `xml/shortcuts.xml` | `app/src/main/res/xml/shortcuts.xml` | The shortcut definition, with an `<extra android:name="shortcut">` so the app can tell which shortcut launched it |
| `values/shortcuts.xml` | `app/src/main/res/values/shortcuts.xml` | Shortcut labels (short / long / disabled message) |
| `drawable/static_shortcut.xml` | `app/src/main/res/drawable/static_shortcut.xml` | A placeholder vector icon |

The consumer copies `resources/` into their project root, replaces `{{BUNDLE_ID}}` in `xml/shortcuts.xml` with their application id, and adds the `<platform name="android">` block from the README to `config.xml`. The `<edit-config>` entry injects the `android.app.shortcuts` meta-data into the launcher activity, which is the part that is easy to get wrong by hand.

### 2. `getIntent()` now consumes the shortcut extras

`ShortcutsPlugin.getIntent()` returns the launching intent, but the activity's intent is sticky: on Android the same `Intent` stays attached to the activity, so every later `getIntent()` call — after a resume, a re-navigation, or a second read on the JS side — kept returning the same `shortcut` extra. The app then re-triggered the shortcut action even though the user never tapped the shortcut again.

After building the result, any extra whose key starts with `shortcut` is removed from the activity intent, so the shortcut is delivered exactly once. The plugin result is sent before the cleanup, so the first caller still gets the full payload, and non-shortcut extras are left untouched.

### 3. Ionic / Angular example

The `@awesome-cordova-plugins/shortcuts-android` wrapper is published (added in [danielsogl/awesome-cordova-plugins#4831](https://github.com/danielsogl/awesome-cordova-plugins/pull/4831)), so the README now shows the idiomatic way to consume both entry points from a single stream:

```ts
merge(from(this.shortcutsAndroid.getIntent()), this.shortcutsAndroid.onNewIntent())
  .pipe(filter((intent: Intent) => intent?.extras?.shortcut))
```

`getIntent()` covers a cold start from a shortcut, `onNewIntent()` covers a shortcut tap while the app is already running — an app needs both, and this is the piece that was previously undocumented.

## Testing

Verified on a Cordova app built from these resource templates (Android 8+):

- The static shortcut is listed on long-press of the launcher icon.
- Cold start from the shortcut: `getIntent()` returns `extras.shortcut === 'static_shortcut'`; a second `getIntent()` in the same session no longer reports it.
- Tapping the shortcut while the app is already running fires `onNewIntent()` with the extra.
- A normal launch from the launcher reports no `shortcut` extra.

To reproduce: copy `resources/` into a Cordova app, add the `<platform name="android">` block from the README to `config.xml`, replace `{{BUNDLE_ID}}`, and build for Android.

## Notes for the reviewer

- Behaviour change: repeat `getIntent()` calls no longer re-report a shortcut. This is the intended fix, but it is a change for anyone who was relying on the previous sticky behaviour, so it may deserve a `CHANGES` entry / version bump — happy to add one in whichever form you prefer.
- The icon is a placeholder vector; consumers are expected to replace it.

MaximBelov committed:
doc: Add example for static shortcuts

MaximBelov committed:
doc: Add example for static shortcuts

MaximBelov commented on src/android/ShortcutsPlugin.java:
```diff
@@ -126,6 +126,14 @@ private void getIntent(CallbackContext callbackContext) throws JSONException  {
         Intent intent = this.cordova.getActivity().getIntent();
         PluginResult result = new PluginResult(PluginResult.Status.OK, buildIntent(intent));
         callbackContext.sendPluginResult(result);
+        Bundle data = intent.getExtras();
```
For handle case
1. Application started from a static shortcut
2. Restart the application via window.location.href = '/'

MaximBelov committed:
doc: Add an Ionic example

MaximBelov committed:
doc: Add an Ionic example

MaximBelov committed:
doc: Link the published Ionic Native wrapper

The wrapper has been released in @awesome-cordova-plugins, so the
pending-PR note is no longer accurate.

avargaskun reviewed (commented):
Thanks — the resource templates and the Ionic docs fill a real gap, and the one-shot `getIntent()` fix is the right idea. A few notes, the most important one being how this interacts with a change that landed on master after this PR was opened.

### Interaction with master (`365e928`)

Since this PR was created, master rewrote `getIntent()`: `pluginInitialize()` now captures the launch intent up front (to defend against plugins like phonegap-nfc wiping it via `setIntent(new Intent())`), and the first `getIntent()` call consumes that saved reference (`launchIntent = null`). Your branch merges cleanly, and the combined method becomes:

```java
Intent intent = (launchIntent != null) ? launchIntent : this.cordova.getActivity().getIntent();
launchIntent = null; // consume so it is not re-processed
PluginResult result = new PluginResult(PluginResult.Status.OK, buildIntent(intent));
callbackContext.sendPluginResult(result);
Bundle data = intent.getExtras();
if (data != null) {
    for (String key : data.keySet()) {
        if (key.startsWith("shortcut")) {
            intent.removeExtra(key);
        }
    }
}
```

Two takeaways from tracing this:

1. **Your fix is *not* made redundant by `launchIntent = null`.** The consume only clears the saved reference; a second `getIntent()` falls back to `activity.getIntent()`, which is normally the *same sticky Intent object*, still carrying the `shortcut` extra. Without your strip, the re-report bug survives on master today. So the change is still needed.
2. **But the merged code only dedupes correctly because of an object-identity coincidence.** On the first call, `intent` is the saved `launchIntent`; stripping it also strips the activity's intent only because they are (usually) the same object. That happens to hold in every scenario I could construct — including the phonegap-nfc case, where the replaced activity intent has no shortcut extra anyway — but the invariant is now implicit and fragile. Please rebase onto master and re-run your cold-start / double-`getIntent()` test against the merged code, since what you verified in September was the pre-`365e928` method.

It would also be worth adding a sentence to the existing `getIntent` section of the README documenting the new one-shot behavior (extras whose key starts with `shortcut` are consumed after the first read) — right now that contract only lives in the PR description.

### CHANGES / version

On your question in the PR notes: yes, please add a `CHANGES` entry and bump to `0.1.3` (`package.json` + `plugin.xml`) — repeat `getIntent()` calls no longer re-reporting the shortcut is a native behavior change and deserves a line, same as the `0.1.2` entry.

Inline comments on the smaller points below.

MaximBelov committed:
fix: Consume only reserved shortcut extras, and consume them explicitly

Addresses review feedback on the static shortcuts PR.

- Narrow the consumed keys from any "shortcut" prefix to the exact key
  "shortcut" plus the reserved "shortcut." namespace, so app-defined
  extras such as shortcuts_enabled are no longer swallowed. The
  namespace is kept because real consumers use keys like
  shortcut.action rather than the bare key.
- Strip the extras from both the Intent that was reported and the
  activity's current Intent, so the one-shot contract no longer
  depends on those two being the same object.
- Document the one-shot behaviour and the reserved keys in the
  getIntent section of the README.
- Template: rename the shortcut id to static_shortcut for consistency
  with the label, extra and drawable, and drop the <categories>
  element, which takes shortcut categories rather than intent
  categories and was ignored by the launcher.
- Ionic example: drop the unused Shortcut import and make the filter
  predicate return a boolean so it compiles under strict TypeScript.
- Bump to 0.1.3 and add the CHANGES entry.

MaximBelov reviewed (commented):
Thanks for the trace — agreed on almost all of it, with one substantive push-back on the key matching. This branch is from September 2024 against `9d7bc27`, so everything I verified back then predates your `getIntent()` rewrite. Rebased and re-verified against the merged method on an emulator; details below.

### Rebase

Rebased onto `365e928`. All three commits replay with no conflicts and the resolved method is exactly what you quoted. (The PR's `rebaseable: false` on the API looks stale — `mergeable` is `true` and `mergeable_state` is `blocked` on the review gate, not on conflicts.)

### The object-identity coincidence

You're right, and I've made the invariant explicit rather than leaving it to `launchIntent` and `activity.getIntent()` being the same object:

```java
private static final String EXTRA_SHORTCUT = "shortcut";
private static final String EXTRA_SHORTCUT_NAMESPACE = "shortcut.";

private void getIntent(CallbackContext callbackContext) throws JSONException  {
    Intent intent = (launchIntent != null) ? launchIntent : this.cordova.getActivity().getIntent();
    launchIntent = null; // consume so it is not re-processed
    PluginResult result = new PluginResult(PluginResult.Status.OK, buildIntent(intent));
    callbackContext.sendPluginResult(result);
    // Consume the shortcut extras so a later getIntent() does not report the same
    // shortcut again. Strip both the intent we just reported and the activity's
    // current intent: they are normally the same object, but another plugin may
    // have swapped the activity's intent via setIntent(), and the one-shot
    // contract should not depend on that identity holding.
    consumeShortcutExtras(intent);
    consumeShortcutExtras(this.cordova.getActivity().getIntent());
}

private void consumeShortcutExtras(Intent intent) {
    if (intent == null) {
        return;
    }
    Bundle data = intent.getExtras();
    if (data == null) {
        return;
    }
    for (String key : data.keySet()) {
        if (EXTRA_SHORTCUT.equals(key) || key.startsWith(EXTRA_SHORTCUT_NAMESPACE)) {
            intent.removeExtra(key);
        }
    }
}
```

`removeExtra` on an absent key is a no-op, so the common same-object case costs nothing, and "no shortcut extra is reachable after the first read" no longer rests on an implicit assumption. `Intent.getExtras()` hands back a copy of the bundle, so removing while iterating its key set is safe.

### Retest against the merged code

Cordova 13 / cordova-android 15.1.0, API 36 emulator, app built from the resource templates exactly as the README describes.

| Case | Result |
| --- | --- |
| Static shortcut listed on launcher long-press | Shown as "Static Shortcut" with the template icon |
| System registration (`dumpsys shortcut`) | `id=static_shortcut`, manifest-declared, intent carries `PersistableBundle[{shortcut=static_shortcut}]`, `categories=null` |
| Cold start by tapping the real shortcut | `getIntent#1` → `extras: {"shortcut":"static_shortcut"}` |
| Second `getIntent()` in the same session | no `extras` at all |
| `getIntent()` after a resume | no `shortcut` extra |
| Shortcut delivered while the app is running | `onNewIntent` → `extras: {"shortcut":"static_shortcut"}` |
| Plain launcher start | no `extras` at all, on both reads |
| Namespaced key + unrelated extras (see the inline reply) | `shortcut.action` consumed after the first read; `shortcuts_enabled` and `shortcutTheAppDefines` untouched |

The `<edit-config>` meta-data injection and all three `<resource-file>` copies landed correctly in the generated project, so the README instructions work as written.

### README

Added a paragraph to the existing *Querying current Intent* section stating the one-shot behaviour and which keys are reserved, so the contract lives in the docs and not only in this PR description.

### CHANGES / version

Bumped `package.json` and `plugin.xml` to `0.1.3`. Two notes:

- There's no `CHANGES` file in the repo — the changelog is the `## CHANGES` section at the bottom of `README.md`, so the entry went there.
- `365e928` has no changelog line either and master is still on `0.1.2`, so a `0.1.3` bump ships it too. I've covered both changes under `v0.1.3` — happy to drop the launch-intent half if you'd rather log that one separately.

One last thing, since it was pending when this PR was opened: the wrapper is published — `@awesome-cordova-plugins/shortcuts-android@9.5.0` — so the README reference is now accurate. I checked the example against its published typings while fixing the nits.

avargaskun reviewed (approved):
Thank you for the contribution — and for the thorough follow-up. The rebase onto `365e928`, the explicit `consumeShortcutExtras()` on both intents, the reserved `shortcut` / `shortcut.*` namespace (good push-back on the exact match — your production use case is exactly the `shortcut.foo` scenario), the README contract, and the emulator test matrix all check out against the diff.

On the changelog question: keeping the launch-intent fix under the combined `v0.1.3` entry is fine — 0.1.3 genuinely ships both changes, no need to split it.
